homeassistant/components/switch/tplink.py

```diff
@@ -0,0 +1,155 @@
+
+def setup_platform(hass, config, add_devices_callback, discovery_info=None):
+    """Setup the TPLink platform in configuration.yaml."""
+    host = config.get(CONF_HOST)
+    name = config.get(CONF_NAME, DEVICE_DEFAULT_NAME)
+
+    add_devices_callback([SmartPlugSwitch(SmartPlug(host),
+                                          name)])
+
+class SmartPlugSwitch(SwitchDevice):
+    """Representation of a TPLink Smart Plug switch."""
+
+    def __init__(self, smartplug, name):
+        """Initialize the switch."""
+        self.smartplug = smartplug
+        self._name = name
+
+    @property
+    def name(self):
+        """Return the name of the Smart Plug, if any."""
+        return self._name
+
+    @property
+    def is_on(self):
+        """Return true if switch is on."""
+        return self.smartplug.state == 'ON'
+
+    def turn_on(self, **kwargs):
+        """Turn the switch on."""
+        self.smartplug.state = 'ON'
+
+    def turn_off(self):
+        """Turn the switch off."""
+        self.smartplug.state = 'OFF'
+
+class SmartPlug(object):
+    """Class to access TPLink Switch.
+    
+    Usage example when used as library:
+    p = SmartPlug("192.168.1.105")
+    # change state of plug
+    p.state = "OFF"
+    p.state = "ON"
+    # query and print current state of plug
+    print(p.state)
+    Note:
+    The library references the same structure as defined for the D-Link Switch
+    """
+
+    def __init__(self, ip):
+        """Create a new SmartPlug instance identified by the IP."""
+        self.ip = ip
+        self.port = 9999
+        self._error_report = False
+
+    @property
+    def state(self):
+        """Get the device state (i.e. ON or OFF)."""
+        response = self.hs100_status()
+        if response is None:
+            return 'unknown'
+        elif response == 0:
+            return "OFF"
+        elif response == 1:
+            return "ON"
+        else:
+            _LOGGER.warning("Unknown state %s returned" % str(response))
+            return 'unknown'
+
+    @state.setter
+    def state(self, value):
+        """Set device state.
+
+        :type value: str
+        :param value: Future state (either ON or OFF)
+        """
+        if value.upper() == 'ON':
+            s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            s.connect((self.ip, self.port))
+            s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            s.connect((self.ip, self.port))
+            on_str = '0000002ad0f281f88bff9af7d5',
+                     'ef94b6c5a0d48bf99cf091e8b7',
+                     'c4b0d1a5c0e2d8a381f286e793',
+                     'f6d4eedfa2dfa2'
+            data = codecs.decode(on_str, 'hex_codec')
+            s.send(data)
+            s.close()
+
+        elif value.upper() == 'OFF':
+            s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            s.connect((self.ip, self.port))
+            off_str = '0000002ad0f281f88bff9af7d5',
```

Comment:
It may be advised to separate the protocol out into a separate Python package, that way we are not requiring a ton of updates to this code to publish new features, and so that the public can make use of the package as well without having to go through Home Assistant.